honeybee_energy/writer.py

```diff
@@ -22,44 +28,368 @@ def generate_idf_string(object_type, values, comments=None):
+    comments = ('name',
+                'surface type',
+                'construction name',
+                'building surface name',
+                'boundary condition object',
+                'view factor to ground',
+                'frame and divider name',
+                'multiplier',
+                'number of vertices',
+                '')
+    return generate_idf_string('FenestrationSurface:Detailed', values, comments)
+
+
+def aperture_to_idf(aperture):
+    """Generate an IDF string representation of an Aperture.
+
+    Note that the resulting string does not include full construction definitions.
+
+    Also note that this does not include any of the shades assigned to the Aperture
+    in the resulting string. To write these objects into a final string, you must
+    loop through the Aperture.shades, and call the to.idf method on each one.
+
+    Args:
+        aperture: A honeyee Aperture for which an IDF representation will be returned.
+    """
+    ap_bc_obj = aperture.boundary_condition.boundary_condition_object if \
+        isinstance(aperture.boundary_condition, Surface) else ''
+    values = (aperture.name,
+              'Window',
+              aperture.properties.energy.construction.name,
+              aperture.parent.name if aperture.has_parent else 'unknown',
+              ap_bc_obj,
+              aperture.boundary_condition.view_factor,
+              '',  # TODO: Implement Frame and Divider objects on WindowConstructions
+              '1',
+              len(aperture.vertices),
+              ',\n '.join('%.3f, %.3f, %.3f' % (v[0], v[1], v[2])
+                          for v in aperture.upper_left_vertices))
+    comments = ('name',
+                'surface type',
+                'construction name',
+                'building surface name',
+                'boundary condition object',
+                'view factor to ground',
+                'frame and divider name',
+                'multiplier',
+                'number of vertices',
+                '')
+    return generate_idf_string('FenestrationSurface:Detailed', values, comments)
+
+
+def shade_to_idf(shade):
+    """Generate an IDF string representation of a Shade.
+
+    Note that the resulting string will possess both the Shading object
+    as well as a ShadingProperty:Reflectance if the Shade's construction
+    is not defaulted.
+
+    Args:
+        shade: A honeyee Shade for which an IDF representation will be returned.
+    """
+    # create the Shading:Detailed IDF string
+    trans_sched = shade.properties.energy.transmittance_schedule.name if \
+        shade.properties.energy.transmittance_schedule is not None else ''
+    if shade.has_parent and not isinstance(shade.parent, Room):
+        if isinstance(shade.parent, Face):
+            base_srf = shade.parent.name
+        else:  # Aperture for parent
+            try:
+                base_srf = shade.parent.parent.name
```

Comment:
In what case this would / should work?

Reply:
Whenever the parent object hosting the `Shade` is an `Aperture` and that `Aperture` has a parent `Face`. This really should be the case for any energy-simulate-able `Model` but might not be the case if someone passes a shade from an orphaned aperture into this method. So the use of `try/except` here is because the `except` case is truly exceptional.